crates/adapters/dydx/src/websocket/handler.rs

```diff
@@ -92,25 +123,452 @@ impl FeedHandler {
             DydxWsMessage::Error(err) => Ok(Some(NautilusWsMessage::Error(err))),
             DydxWsMessage::Reconnected => Ok(Some(NautilusWsMessage::Reconnected)),
             DydxWsMessage::Pong => Ok(None),
             DydxWsMessage::Raw(_) => Ok(None),
         }
     }
+
+    fn handle_channel_data(
+        &self,
+        data: DydxWsChannelDataMsg,
+    ) -> DydxWsResult<Option<NautilusWsMessage>> {
+        match data.channel {
+            DydxWsChannel::Trades => self.parse_trades(&data),
+            DydxWsChannel::Orderbook => self.parse_orderbook(&data, false),
+            DydxWsChannel::Candles => self.parse_candles(&data),
+            DydxWsChannel::Markets => self.parse_markets(&data),
+            DydxWsChannel::Subaccounts => {
+                // TODO: Parse subaccount updates (orders, fills, positions)
+                tracing::debug!("Subaccount channel_data not yet implemented");
+                Ok(None)
+            }
+            DydxWsChannel::BlockHeight => {
+                tracing::debug!("Block height update received");
+                Ok(None)
+            }
+        }
+    }
+
+    fn handle_channel_batch_data(
+        &self,
+        data: DydxWsChannelBatchDataMsg,
+    ) -> DydxWsResult<Option<NautilusWsMessage>> {
+        match data.channel {
+            DydxWsChannel::Orderbook => self.parse_orderbook_batch(&data),
+            _ => {
+                tracing::warn!("Unexpected batch data for channel: {:?}", data.channel);
+                Ok(None)
+            }
+        }
+    }
+
+    fn parse_trades(&self, data: &DydxWsChannelDataMsg) -> DydxWsResult<Option<NautilusWsMessage>> {
+        let symbol = data
+            .id
+            .as_ref()
+            .ok_or_else(|| DydxWsError::Parse("Missing id for trades channel".into()))?;
+
+        let instrument_id = self.parse_instrument_id(symbol)?;
+        let instrument = self.get_instrument(&instrument_id)?;
+
+        let contents: DydxTradeContents = serde_json::from_value(data.contents.clone())
+            .map_err(|e| DydxWsError::Parse(format!("Failed to parse trade contents: {e}")))?;
+
+        let mut ticks = Vec::new();
+        let ts_init = nautilus_core::time::get_atomic_clock_realtime().get_time_ns();
+
+        for trade in contents.trades {
+            let aggressor_side = match trade.side {
+                DydxOrderSide::Buy => AggressorSide::Buyer,
+                DydxOrderSide::Sell => AggressorSide::Seller,
+            };
+
+            let price = Decimal::from_str(&trade.price)
+                .map_err(|e| DydxWsError::Parse(format!("Failed to parse trade price: {e}")))?;
+
+            let size = Decimal::from_str(&trade.size)
+                .map_err(|e| DydxWsError::Parse(format!("Failed to parse trade size: {e}")))?;
+
+            let tick = TradeTick::new(
+                instrument_id,
+                Price::new(price.try_into().unwrap(), instrument.price_precision()),
```

Comment:
Lets change these unwraps, there are methods on most value types which can return results then `?`, then propagate up, or map into this domain error.

Reply:
Done, thx